81-question-mark-chrysalis/technical_specification.md

```diff
@@ -0,0 +1,233 @@
+
+We enforce this stronger unique-path property because "diamond" shapes indicate unnecessary redundancy that should be avoided, and because it eliminates any ambiguity in how data are transformed.
+
+This graph structure is used to validate the transformation configuration and to determine the processing order of schemas during model derivation and configuration changes.
+
+#### Transformation Configuration Validation
+
+Transformation configuration validation includes:
+
+1. Verify that all workflows and EMIMs (i.e., models with is_ingest = true) referenced by the routes exist in the configuration.
+2. Ensure that EMIMs do not appear as the output models of any routes.
+3. Validate schemas using the SchemaPack library and workflows using the metldata library.
+4. Confirm that the graph meets the unique-path requirement and is therefore acyclic.
+5. Compute a topological order of the graph’s nodes—e.g. using Kahn’s algorithm—so that each model is processed only after all its dependencies when visiting the models in that order.
+6. Update the order field of each model based on the computed topological order.
+
+
+#### User Journeys: Model Derivation
+
+When manually triggered—or when the configuration changes—the service derives the output schemas for all routes.
+
+1. Validate the transformation configuration as described above.
+   1. The configuration is rejected if validation fails.
+   2. In case of rejection, the previous valid configuration remains active.
+2. Traverse the transformation graph starting from the EMIMs, following the topological order. For each route:
+   1. Use metldata to run the workflow identified by `workflow_name` on the input model’s schema referenced by `input_model_name` and compute the derived schema.
+   2. Update the output model’s schema accordingly.
+3. If any errors or conflicts occur, abort the operation and report them.
+
+
+#### User Journeys: Service Consumer Transforms An Original AnnotatedEMPack
+
+The transformation operation is triggered when the service receives either an “original AnnotatedEMPack upsert” event or a “re-transform AnnotatedEMPack” event.
+
+1. Build the dirty map: Query the AnnotatedEMPacks collection for all data where `original_id` matches the incoming original data's ID. Extract each data's ID and model name, then create a mapping from model names to data IDs. This "dirty map" tracks data that need to be either re-created or deleted.
+
+2. Initialize the transformed map: Create a mapping from model names to data. This map will accumulate all data generated during this transformation operation. Initialize it with a single entry: the original model name mapped to the incoming original data.
+
+3. Traverse the transformation graph: Starting from the EMIM and following the topological order, perform the following for each model:
```

Comment:
Following the topological order means skipping over entries in that order that don't match any data accumulated so far?

Reply:
I probably still don't understand the question or which point is still unclear in the algorithm. To me it seems well-defined and unambiguous. Here in step 3 for the traversal we only look at the current model graph, not at any data.